### DISCUSSION TITLE
Expose Authentication/Login API

### DISCUSSION THREAD

**Discussion by mffap** [zitadel/zitadel#4193] [Ideas]
Some users asked for more flexibility on the login screen.

ZITADEL ships by default with a hosted login, which is optimized and hardened for security by default. We are working to improve the flexibility of the hosted login in two ways: 
- More customization: https://github.com/zitadel/zitadel/issues/3987
- We aim to improve [Actions](https://docs.zitadel.com/docs/concepts/features/actions) in the future to customize the login process and layout

For native apps we recommend to use the embedded web-view of the OS to pull up the login screen and redirect back to the app. This is the recommended default with OpenID Connect. You can try this with our [Flutter](https://docs.zitadel.com/docs/examples/login/flutter) example. 

Despite these changes and future improvements, there might still remain cases where users want to implement a fully custom login process and user interface. For those cases it might be beneficial to expose the Authentication/Login API. 

Use cases might include: 
- Bring-your-own login page
- Embedded login with an app (eg, games, vr, ...)

Further considerations: 
- We should provide a component library for web
- Disabling the hosted login and enabling this should an option with a warning (similar to "Dev Mode" in OIDC configuration"), reserved for only management roles

--- 
_Edit (March 22):_ Implementation is being tracked in https://github.com/zitadel/zitadel/issues/5015

_Edit (June 28):__ Checkout the following resources for documentation and feedback:
- [Documentation](https://zitadel.com/docs/guides/integrate/login-ui)
- https://github.com/zitadel/zitadel/discussions/5922
- https://github.com/zitadel/zitadel/discussions/5875

**Comment by MatthewAry** [comment]
I am a fan of this. In order to have complete end-to-end control of customer/user experiences this would be necessary. Making it so this can be handled with relative ease could be done by providing a drop-in package for (in this example) browser front-ends that:
- Has a constructor that accepts configuration parameters
- Constructor returns an object that can be used to manage authentication on the front-end
- Object contains authentication lifecycle methods, events, and hooks
  - Password reset request methods
  - Successful authentication event interceptor (allows for some operation to happen before redirect takes place)
  - i18n support for error/success messages that might be returned
  - FIDO and Passwordless Authentication Support
  - 2 Factor Authentication Support
  - Hooks to pass in tokens and redirect parameters from the URL (along with an optional mechanism to get such information without dev intervention)
- Compatibility with SSR based front-ends
- Compatibility with SPA based front-ends

The expectation is that this package would be used exclusively to allow for the creation of customized authentication experiences and nothing more than that. People would authenticate and then be redirected to the appropriate application. Users may also want to use their customized authentication application to manage account registrations, legal agreements, surveys, and other setup flows.

I probably could write a lot more about this but this is what I have had come to mind.

**Reply by fforootd** [reply]
Ok, so let me dump what we thought so far:

- Create an easy to use API that allows to trigger either an auth (login) or register flow
  - Allow register and login with local and as well federated users
  - Extend the API flows with the help of Actions to allow for push style integrations (like webhooks but more flexible) (:construction: this is already in progress https://github.com/zitadel/zitadel/issues/4265)
  - Allow users to change the tokens (:construction: this is already in progress https://github.com/zitadel/zitadel/issues/4265)
  - We need to discuss how passwordless work in this case (since FIDO2 verifiers the domain)
- We need to define some kind of protection for this API
  - URL / Domain whitelist
  - Something to authenticate the "client"
- Use cases that could benefit from this
  - People who want to build their own guis for login and register (in whatever tech)

**Comment by fdelucchijr** [comment]
This idea should be a must have. 

UI libraries are pretty much appreciated but in most cases isn't an issue, as (in case IaaS doesn't exist) the developers would be ready to program the components themselves.

The actual API for Auth is the very thing that needs to be solved. None of the Admin/User endpoints offer a way to authenticate the user.

I think the future way is to expose a public api for Login/Register, but in the short term Management endpoint should have a call for authenticate the user with his credentials (there's already a call to create a human user), as allow programmers to handle the message at our api and make the necessary changes to integrate Zitadel without the friction WebView adds in mobile apps (for example).

**Reply by fforootd** [reply]
> I think the future way is to expose a public api for Login/Register, but in the short term Management endpoint should have a call for authenticate the user with his credentials (there's already a call to create a human user), as allow programmers to handle the message at our api and make the necessary changes to integrate Zitadel without the friction WebView adds in mobile apps (for example).

Thanks for this input. We are at the moment exploring how such an API for Login/Register could be structured.

**Comment by cskeppstedt** [comment]
Hello! Just wanted to document a perhaps slightly different use-case that depends on this kind of API.

I'm working on a project that uses a 3rd party app to authenticate our end users (called [BankID](https://www.bankid.com/en/utvecklare/guider/teknisk-integrationsguide)). It's the de facto way to prove your identity digitally in Sweden, and a common method to sign in on official government/tax authority web apps, as well as commercial web apps. This is roughly how we use it today (the IDP being Keycloak at the moment):

![swimlanes-e518b3e9e1d974ff1491c0acb62c3d26](https://user-images.githubusercontent.com/569742/196812900-cc942177-ecec-4c98-aaef-d7f628f00d70.png)

The registration flow is identical, but of course uses the create-user API endpoint instead of sign-in. When a user registers, we derive a username and password from the unique ID we get from the 3rd party response by hashing it with a unique salt for that particular user.

When the user wants to sign in,  the 3rd party gives us the same unique ID for that user, we will then derive the same hashed username and password again, and thus we can sign in the correct user in Keycloak to get an access-token that the app user can use  towards our backend API.

In this flow, the user never interacts directly with the IDP service (Keycloak currently), but only the main App and the 3rd party app. The end user is never aware of a username or password - they simply register/signin by using the 3rd party app (BankID).

I was very happy when I found out about Zitadel, it solves many issues we have with Keycloak, and just checks so many boxes for us (the org. architecture alone saves us a lot of custom implementation) – but to make the switch we need support for this kind of flow. Anyways, keep up the good work!

**Reply by fforootd** [reply]
Thank you for the kind words and your inputs.

We definitely will invest time into a login/register API that could allow you case.

Since you already use KC, mind me asking what apis you use there and if you like the design 😁

**Reply by cskeppstedt** [reply]
Glad to hear it, can't wait :)

Sure no problem; our backend is written in Typescript, and we're using the official [keycloak-admin-client](https://npmjs.com/package/@keycloak/keycloak-admin-client). Afaik it just wraps calls to their REST API. 90% of our usage is the [Users](https://www.keycloak.org/docs-api/11.0/rest-api/index.html#_users_resource)-resource; either getting a single user by id, or searching for users by name/email. We sign in users with username & password as mentioned above by making a POST to the `token` endpoint with `grant_type=password` (`/realms/{realm}/protocol/openid-connect/token`).

Honestly we haven't interacted that much with the keycloak API, but the parts I've looked at seems OK to me design wise. Our main pain points with keycloak are:

1. Upgrading to newer versions. When we started our journey the major version was 11.0, and the current version is 19.0. I've looked at upgrading many times but I always decide that it seems like more work than it's worth. The rest of our infra is modern and serverless, but keycloak stands out as a bit old school here with needing to run commands inside the container (`kcadm.sh`). There are too many manual steps. It's not cloud native you could say:)

2. Lack of sorting and filtering on custom attributes. We want to have the list of users paginated and sorted by created, descending, so we can see the most recent users first. This is _still_ not supported, even though there have been 8 major versions since we started. The workaround is to either fetch all (thousands) of users, and sort them in the consuming application, or maintain some sort of index in parallell that supports sorting. We're doing the former today, and when that doesn't scale anymore we'd prefer to not having to do the latter. Btw AWS cognito doesn't support sorting either. I think it's all a bit odd, how is that not a default feature 🤷

**Reply by fdelucchijr** [reply]
> Btw AWS cognito doesn't support sorting either. I think it's all a bit odd, how is that not a default feature 🤷 

Is interesting, Ory Kratos also doesn't support it. As I remember Auth0 either (I might be wrong) but it lets you sync db with your own infrastructure for you to implement the searching, indexing, etc (That adds a lot of overhead and work of the development team to implement in a secure/stable way). 
Ory thinks into develop a kind-of GraphQL, but that seems gonna take a lot more time to be finally released.

As I see, currently Identity-aaS doesn't implement a flexible system to store **User Profiles** and is more oriented to store actual identity's.

PD: I understand "Profile" as data that will be shown to other users, for example your alias, birthday, gender, etc. And "Identity" as who really are behind the screen, how it logs in and what permissions possess.

Currently our IaaS only stores the email, phone and password as they are strictly private for the user and are central parts of the identity, and the profile part is managed by us as we can't do search's, pagination, etc based in profile traits with or IaaS. (Keep in mind that our app has a social network built in)

There are many points related to that will make this comment larger. Maybe a forum in the discord server or a discussion in the repo will be better to address this thread.

**Reply by fforootd** [reply]
Hm interesting thoughts.

Since we rely on normalized views and SQL it should not be to tricky to allow for sorting options there.

We also talked about allow to search through the users metadata, however we did not yet tackle this one.

@fdelucchijr I would love to discuss this further, for us either discord or here is fine 😁 as long as it is in public and searchable (thats why we do not use slack)

**Reply by JesperWe** [reply]
@cskeppstedt I am currently looking at what IAM tool to recommend in a project that will require BankId as IdP. Just curious how far you got in this? Some, but not all, bits seem to have been implemented in Zitadel already. Are things at a state where direct BankId integration is feasible?

**Reply by scheibling** [reply]
@JesperWe I'm not cskeppstedt, but we're currently finishing up a piece of software to translate BankID to Oauth2.0 flows, which would mean it'd be a lot easier to integrate into all kinds of IDPs. If you haven't developed your own solution yet we might be able to help

**Comment by Davincible** [comment]
Since this is a discussion, is there any issue open to track the progress on this?

**Reply by mffap** [reply]
@hifabienne can you link the implementation issues for reference, please?

**Reply by hifabienne** [reply]
I have just created the issue to this: https://github.com/zitadel/zitadel/issues/5015
It also includes the registration api.

**Reply by mffap** [reply]
Checkout the following resources for documentation and feedback:
- [Documentation](https://zitadel.com/docs/guides/integrate/login-ui)
- https://github.com/zitadel/zitadel/discussions/5922
- https://github.com/zitadel/zitadel/discussions/5875

**Comment by Brick-Logic** [comment]
Any news on this? I would like to use this with .NET MAUI Blazor App with a custom login page.

**Reply by hifabienne** [reply]
The session API is published in an beta state, you can find the docs here: https://zitadel.com/docs/apis/resources/session_service
Or some more guides on how to build your own login ui here: https://zitadel.com/docs/guides/integrate/login-ui
